packages/unified-latex-to-pretext/libs/pre-conversion-subs/break-on-boundaries.ts

```diff
@@ -0,0 +1,146 @@
+import { env, arg } from "@unified-latex/unified-latex-builder";
+import * as Ast from "@unified-latex/unified-latex-types";
+import { getNamedArgsContent } from "@unified-latex/unified-latex-util-arguments";
+import {
+    anyEnvironment,
+    anyMacro,
+    match,
+} from "@unified-latex/unified-latex-util-match";
+import { printRaw } from "@unified-latex/unified-latex-util-print-raw";
+import { replaceNode } from "@unified-latex/unified-latex-util-replace";
+import {
+    splitOnMacro,
+    unsplitOnMacro,
+} from "@unified-latex/unified-latex-util-split";
+import { visit } from "@unified-latex/unified-latex-util-visit";
+
+/**
+ * All the divisions, where each item is {division macro, mapped environment}.
+ * Note that this is ordered from the "largest" division to the "smallest" division.
+ */
+const divisions: { division: string; mappedEnviron: string }[] = [
+    { division: "part", mappedEnviron: "_part" },
+    { division: "chapter", mappedEnviron: "_chapter" },
+    { division: "section", mappedEnviron: "_section" },
+    { division: "subsection", mappedEnviron: "_subsection" },
+    { division: "subsubsection", mappedEnviron: "_subsubsection" },
+    { division: "paragraph", mappedEnviron: "_paragraph" },
+    { division: "subparagraph", mappedEnviron: "_subparagraph" },
+];
+
+/**
+ * Breaks up division macros into environments. Returns a list of warning messages
+ * for any groups that were removed.
+ */
+export function breakOnBoundaries(ast: Ast.Ast): { messages: string[] } {
+    // messages for any groups removed
+    const messagesLst: { messages: string[] } = { messages: [] };
+
+    // check if a macro is a division macro
+    const isDivision = match.createMacroMatcher(
+        divisions.map((x) => x.division)
+    );
+
+    visit(ast, (node, info) => {
+        // needs to be an environment, root, or group node
+        if (
+            !(
+                anyEnvironment(node) ||
+                node.type === "root" ||
+                match.group(node)
+            ) ||
+            // skip math mode
+            info.context.hasMathModeAncestor
+        ) {
+            return;
+        }
+        // if it's an environment, make sure it isn't a newly created one
+        else if (
+            anyEnvironment(node) &&
+            divisions.map((x) => x.mappedEnviron).includes(node.env)
+        ) {
+            return;
+        }
+
+        // add message for groups to be removed that contain a division as an immediate child
+        if (match.group(node) && isDivision(node.content[0])) {
```

Comment:
We want to know if _any_ immediate child is a division, not just the first one.

Reply:
I think it is best to do two passes. The first pass only removes groups. The second pass doesn't care about groups and just breaks up based on division.